docs/en/DevelopingUFSExtensions.md

```diff
@@ -0,0 +1,138 @@
+/alluxio-extensions/tree/master/underfs/template) repository. In the rest of this section we
+describe the steps involved in writing a new under storage extension. The sample project, called
+`DummyUnderFileSystem`, uses maven as the build and dependency management tool, and forwards all
+operations to a local filesystem.
+
+## Implement the Under Storage Interface
+
+Refer to [integrating under storage systems](Integrating-Under-Storage-Systems.html) for
+additional instructions for developing and testing an under storage module.
+
+Step 1: Implement the interface `UnderFileSystem`
+
+The `UnderFileSystem` interface is defined in the module `org.alluxio:alluxio-core-common`. Choose
+to extend either `BaseUnderFileSystem` or `ObjectUnderFileSystem` to implement the `UnderFileSystem`
+interface. `ObjectUnderFileSystem` is suitable for connecting to object storage and abstracts away
+mapping file system operations to an object store.
+
+```java
+public class DummyUnderFileSystem extends BaseUnderFileSystem {
+	// Implement filesystem operations
+	...
+}
+```
+
+or,
+
+```java
+public class DummyUnderFileSystem extends ObjectUnderFileSystem {
+	// Implement object store operations 
+	...
+}
+```
+
+Step 2: Implement the interface `UnderFileSystemFactory`
+
+The under storage factory determines when an interface of the implemented `UnderFileSystem` is
+instantiated.
+
+```java
+public class DummyUnderFileSystemFactory implements UnderFileSystemFactory {
+	...
+	public boolean supportsPath(String path) {
+		// Choose which schemes to support, e.g., dummy://
+	}
+}
+```
+
+## Declare the Service
+
+Create a file at `src/main/resources/META_INF/services/alluxio.underfs.UnderFileSystemFactory`
+advertising the implemented `UnderFileSystemFactory` to the ServiceLoader.
+
+```
+alluxio.underfs.dummy.DummyUnderFileSystemFactory
+```
+
+## Build
+
+Include all transitive dependencies of the extension project in the built JAR using either
+`maven-shade-plugin` or `maven-assembly`.
+
+In addition, to avoid collisions specify scope for the dependency `alluxio-core-common` as
+`provided`. The maven definition would look like:
+
+```xml
+<dependencies>
+    <!-- Core Alluxio dependencies -->
+    <dependency>
+      <groupId>org.alluxio</groupId>
+      <artifactId>alluxio-core-common</artifactId>
+      <scope>provided</scope>
+    </dependency>
+    ...
+</dependencies>
+```
+
+## Test
+
+Extend `AbstractUnderFileSystemContractTest` to test that the defined `UnderFileSystem` adheres to
+the contract between Alluxio and an under storage module. Look at the reference implementation to
+include parameters such as the working directory for the test.
+
+```java 
+public final class DummyUnderFileSystemContractTest extends AbstractUnderFileSystemContractTest {
+    ...
+} 
+```
+
+Congratulations!. You have developed a new under storage extension to Alluxio. Let the community
```

Comment:
!.

Reply:
what do you mean?

Reply:
`Congratulations!.` -- unnecessary period